src/Moq/Linq/MockSetupsBuilder.cs

```diff
@@ -212,24 +191,89 @@ private static Expression ConvertToSetupProperty(Expression targetObject, Expres
+					v.MockObject),
+				// expression:
+				Expression.Lambda(
+					rewrittenLeft,
+					v.MockObjectParameter),
+				// value:
+				Expression.Convert(right, typeof(object)));  // explicit boxing operation required for value types
+		}
+
+		/// <summary>
+		///   Locates the root mock object in a setup expression (which is usually, but not always, a <see cref="ParameterExpression"/>),
+		///   stores a reference to it, and finally replaces it with a new <see cref="ParameterExpression"/>.
+		/// </summary>
+		private sealed class ReplaceMockObjectWithParameter : ExpressionVisitor
+		{
+			private Expression mockObject;
+			private ParameterExpression mockObjectParameter;
+
+			public Expression MockObject => this.mockObject;
 
-			var returnsMethod = typeof(IReturns<,>)
-				.MakeGenericType(sourceType, returnType)
-				.GetMethod("Returns", new[] { returnType });
+			public ParameterExpression MockObjectParameter => this.mockObjectParameter;
 
-			if (right is ConstantExpression constExpr && constExpr.Value == null)
+			protected override Expression VisitMember(MemberExpression node)
 			{
-				right = Expression.Constant(null, left.Type);
+				if (node.Expression is ParameterExpression pe && pe.Type.IsDefined(typeof(CompilerGeneratedAttribute)))
+				{
+					// In LINQ query expressions with more than one `from` clause such as:
+					//
+					//   from a in Mocks.Of<A>()
+					//   from b in Mocks.Of<B>()
+					//   where ...
+					//
+					// The `.Where()` operator will have a parameter `p` that is neither of type `A` nor `B`,
+					// but of a compiler-generated anonymous type (essentially that of `new { a, b }`).
+					//
+					// In such cases, the mock object is not referenced by `p` itself, but rather by any
+					// member of `p`, that is, `p.a` or `p.b`:
+					mockObject = node;
+					mockObjectParameter = Expression.Parameter(node.Type, pe.Name);
+					return mockObjectParameter;
+				}
+				else
+				{
+					return base.VisitMember(node);
+				}
 			}
 
-			return Expression.NotEqual(
-				Expression.Call(VisitFluent(left), returnsMethod, right),
-				Expression.Constant(null));
+			protected override Expression VisitParameter(ParameterExpression node)
+			{
+				// Regular case: The parameter `p` directly references the mock object:
+				mockObject = node;
+				mockObjectParameter = Expression.Parameter(node.Type, node.Name);
+				return mockObjectParameter;
+			}
+
+			protected override Expression VisitUnary(UnaryExpression node)
+			{
+				// An expression may contain nested expressions, such as:
+				//
+				//     m => m.GetObjectById(It.IsAny<int>(id => id > 0))
+				//          ^                                   ^^
+				//   We're only interested            ...but not in this one!
+				//   in this parameter...
+				//
+				// Therefore, for the nested (quoted) expression, we short-circuit:
+				return node.NodeType == ExpressionType.Quote ? node : base.VisitUnary(node);
```

Comment:
It *may* be safer to replace this guard with a different algorithm:

1. Implement a new visitor that travels along the "left spine" of an expression until it arrives at the root `ParameterExpression`, then it stores a reference to it.

2. Run that visitor on the `left` expression and make the identified `ParameterExpression` available to the present visitor.

2. Add checks in the present visitor so it only transforms expressions involving the same `ParameterExpression` that was identified in the earlier steps.